rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,221 @@
+
+We will continue to use the new development and deployment pipeline created for the trials. Much of the functionality is defined in [a new `Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb), which works like this:
+
+1. **Integration**
+
+   1. Deployment runs.
+   1. [**new**] Wait for the Smoke tests for the app to pass [[1](https://github.com/alphagov/govuk-puppet/pull/10387)].
+   1. [**new**] Remotely trigger the same deployment in Staging.
+
+1. **Staging**
+
+   1. [**new**] Repeat as for Integration, but to Production.
+
+We propose to apply this automation across our apps. Given the scale of the change, we need to think carefully about how we can make it consistently safe and secure, before we enable CD for an app.
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
+
+#### Check: Internal repo tests pass (i.e. features are tested in detail)
+
+For the purpose of automating deployments:
+
+- We will use code coverage as our indicator of previous, thorough testing. While an app may have a [rigorous testing strategy](https://github.com/alphagov/content-publisher/blob/69bed3535de1e65bda9d5909a3bc56a319e46576/docs/testing-strategy.md), it's hard to know if this has been applied consistently as the app has evolved.
+
+- We will require a minimum level of coverage that is realistic for all of our apps. We will not pursue this once CD is enabled, as doing so could encourage [a lack of rigour in the way we write our tests](https://jeroenmols.com/blog/2017/11/28/coveragproblem/).
+
+- We will not check coverage of periphery code e.g. CSS, JavaScript. This is because it has a low footprint in our codebases, and it's hard to measure coverage for: the frameworks we use don't support it.
```

Comment:
I'm not sure what "the frameworks we use don't support it" means, as we do have JS tests in our frontends. Do you mean specifically that our _coverage_ tools doesn't support JS code coverage? Cc @alex-ju

Reply:
@bilbof yes, there's no change there. The intent of this caveat is to say that we won't mandate the existence or volume of testing for periphery code as part of enabling CD for an app.

One way to be more rigorous with "periphery" and "low footprint" would be to say it should be non-declarative and constitute no more than X% of the codebase <- getting a bit complicated!

But it does worry me that JavaScript constitutes 17% of the code in Finder Frontend, for example.

This is where we may have to make a tough decision: I think JS is really the only code we're concerned about, but as far as I know, there's no way to measure coverage for it due to the test framework we use (jasmine-rails). While we could start down the road of rewriting it in (frankly) a better framework, my suggestion is that we just don't have enough bespoke JS code in each app to justify making coverage of it a criteria for enabling CD.

This is where I'd like some help from everyone:

- @alex-ju and @kevindew have better knowledge about our JS testing woes than me. Am I right that: (a) we're heavily invested in jasmine-rails; and (b) there's no way to get coverage stats for it?

- @chao-xian does that clarify this for you? If so, I can try to suggest some different words for the RFC. Also, do you know of instances where we've deployed broken JS code?

Reply:
@benthorner yes, whilst we still make use of asset-pipeline we need a JS test suite that works with it, our options are quite limited in that regard hence Jasmine. I'm not aware of it being easy to get Jasmine coverage - a quick google seems to show that you end up wrapped in karma for that, which likely pulls us out of the Ruby asset-pipeline world.

Reply:
@benthorner It does - thanks. Static is an example where the cookies code did have bugs going out iirc. Maybe increased coverage for JS in specific apps might be a thing to consider including Finder Fronntend.

Reply:
Thanks for your thoughts @kevindew and @chao-xian. I had a look at the cookie bug you mentioned, which I believe refers to [this PR](https://github.com/alphagov/static/pull/2258), which doesn't seem to have any tests around it <- a bad sign!

Looking at Static got me interested in the general presence of JS across our apps. I managed to hack together a way of getting JS code coverage stats, but it's definitely not something we'd want to do normally! It's very tedious to apply the method, but I've done it for all apps where JS is 5%+ of their codebase (according to GitHub). Otherwise, I just checked if there were at least some JS tests. Here's what I found:

- Collections - 7.9% - 71% coverage [[1](https://github.com/alphagov/collections/pull/2030)]
- Collections Publisher - 4.9% - has some tests [[1](https://github.com/alphagov/collections-publisher/tree/master/spec/javascripts)]
- Content Publisher - 9.8% - 69% coverage [[1](https://github.com/alphagov/content-publisher/pull/2186)]
- ⚠ Content Tagger - 3.7% - evidence of tests [[1](https://github.com/alphagov/content-tagger/blob/b385b91b3f512257ee6f8061c34c99c89fbfe845/spec/features/tag_a_page_spec.rb#L77)]
- Email Alert Frontend - 2.5% - has some tests [[1](https://github.com/alphagov/email-alert-frontend/tree/master/spec/javascripts)]
- ⚠ Feedback - 2.5% - evidence of tests [[1](https://github.com/alphagov/feedback/blob/dbb3f16b7505ff87b62301c7d397713442cd8a6d/spec/controllers/contact/govuk/problem_reports_controller_spec.rb#L99)]
- Frontend - 6.5% - 93% coverage [[1](https://github.com/alphagov/frontend/pull/2499)]
- Finder Frontend - 14.5% - 90% coverage [[1](https://github.com/alphagov/finder-frontend/pull/2264)]
- Government Frontend - 3.6% - has some tests [[1](https://github.com/alphagov/government-frontend/tree/master/spec/javascripts)]
- ⚠ Licence Finder - 7.4% - no tests!!
- Manuals Frontend - 19.6% - 95% coverage [[1](https://github.com/alphagov/manuals-frontend/pull/991)]
- Manuals Publisher - 1.2% - has some tests [[1](https://github.com/alphagov/manuals-publisher/tree/master/spec/javascripts)]
- Maslow - 2.1% - has some tests [[1](https://github.com/alphagov/maslow/tree/master/spec/javascripts)]
- Publisher - 8.5% - 61% coverage [[1](https://github.com/alphagov/publisher/pull/1361)]
- ⚠ Search Admin - 1.3% - no tests!!
- Service Manual Frontend - 29.5% - 59% coverage [[1](https://github.com/alphagov/service-manual-frontend/pull/795)]
- Signon - 1.1% - has some tests [[1](https://github.com/alphagov/signon/tree/master/spec/javascripts)]
- Service Manual Publisher - 7.5% - 50% coverage [[1](https://github.com/alphagov/service-manual-publisher/pull/862)]
- ⚠ Specialist Publisher - 1.2% - no tests!!
- ⚠ Static - 70.6% - ?? (technique didn't work, too many unresolvable test errors)
- Transition - 5% - 88% coverage [[1](https://github.com/alphagov/transition/pull/1012)]
- Travel Advice Publisher - 2.3% - has some tests [[1](https://github.com/alphagov/travel-advice-publisher/tree/master/spec/javascripts)]
- Whitehall - 3.8% - has some tests [[1](https://github.com/alphagov/whitehall/tree/master/test/javascripts)]

My opinion: not great, but not terrible either. One thing I'd like to re-emphasise is that this RFC is primarily about the safety of enabling continuous deployment our apps. With that in mind, I'd like to propose a compromise:

- Replace "periphery code" with JavaScript. I don't think we have any specific concerns around "CSS testing" (there is no such thing, outside of all the kinds of test we have at the moment).

- Require that every app with JS code has at least one JS test. This is because a lack of any tests for the language sets a bad precedent. Feature testing is not a reasonable substitute for actual JS tests.

- Require that Static has 95%+ JS coverage.

@chao-xian does that sound reasonable to you? I'd also like to get thoughts from @alex-ju 👋.

Reply:
TL;DR – I'm happy with the content of this RFC, maybe referring to JavaScript and CSS as 'periphery code' is indeed not a great idea as some people may feel offended, but otherwise all good.

Thoughts on this section:

> We will not check coverage of periphery code e.g. CSS, JavaScript. [...]

One thing to add is that given we use progressive enhancement at GDS, feature tests are meant to cover both scenarios: with JavaScript disable and enabled. This will ensure the main user journeys are covered with tests regardless of the level of UI enhancement done with JavaScript.

> [...] This is because it has a low footprint in our codebases, [...]

As an aside, I would argue that the recent designs ask for more and more JavaScript, which although is against our principles of keeping things simple it's a reality and it's very likely that the footprint will increase over time. This being said I would use the progressive enhancement approach is a stronger argument than the footprint size.

> [..] and it's hard to measure coverage for: the frameworks we use don't support it.

The fact that it's harder to measure JavaScript coverage is true, as the current setup doesn't allow us to do this with ease, but debatable: We can potentially use Instanbul and Jest (which relies on the same Istanbul instrumenter loader for coverage), we have Yarn and use npm packages in a few projects already (govuk_publishing_components, content-publisher, finder-frontend to name a couple of them), but we're only halfway through in having a proper setup and this requires time, mandate and consensus on the approach.

Following my hopefully relevant comments, I would be very keen on having a coverage threshold for our JavaScript tests, but I'm afraid we're not there yet with our setup.

> Require that Static has 95%+ JS coverage.

Not sure what this rule is meant to support in the current context where JavaScript is being moved from `static` to `govuk_publishing_components` and most of the JavaScript code in `static` is used for analytics and is mainly spaghetti – and not in a tasty, high-in-carbs way 🍝 . I can argue that we should aim for all projects to have 95% JavaScript coverage, but that doesn't mean people will jump into adding tests for legacy code that is not even modular.

Reply:
Thanks for your thoughts @alex-ju! I'm glad we've got some FE perspective on this, and I do indeed intend to reword this caveat. And yes, while we continue to use Rails+JS (without Webpacker), there is no general, simple approach to measuring coverage (that I can find). This is a good point to emphasise a couple of things:

- The safety criteria in this RFC are just that: a set of rules to help prevent us doing something reckless. While we are committing to automatic deployments, we should prioritise this against alternatives, like replacing Static.

- The safety criteria need to be measurable. While we should be testing code with/out progressive enhancement, there's no way to measure this. The fact that we have some tests for progressive enhancement is reassuring, but it's still unclear just how much JS in a repo is actually tested, without adopting some kind of measure for it.

I think we're basically in agreement, though. What do you think about the following to replace the current text:

- We will require some evidence of JavaScript testing, but we will not go so far as to check the coverage. This is because of its relatively low footprint in our codebases, the difficulty of measuring coverage for it [[1](https://github.com/alphagov/finder-frontend/pull/2264)], and the fact that we often cover it implicitly with Ruby tests, as part of our UI approach of progressive enhancement. An exception to the need for coverage is [Static](https://github.com/alphagov/static), for which JavaScript constitutes 70% of the codebase.

I'll then adjust the criteria accordingly. How does that sound? Have I captured the points you made OK?

Reply:
(It would also be helpful to have a reference to any guidance we have on progressive enhancement.)

Reply:
Service Manual – [Building a resilient frontend using progressive enhancement](Building a resilient frontend using progressive enhancement
). We also have it in the [GDS Way](https://gds-way.cloudapps.digital/manuals/programming-languages/html.html#content).

Reply:
The proposed phrasing sounds excellent, Ben. Happy to approve this PR, but not sure if that has any weight to be honest.

Reply:
Hopefully it will get through! I'll invite people to approve after Senior Tech have (they make the ultimate decision on these things).

I've added a commit to put this into the main text: https://github.com/alphagov/govuk-rfcs/pull/128/commits/3469831899f39d23de30195e111b44ffdf5e5563

I'll resolve this thread, but feel free to comment if you think there's more to discuss.